Enhancing tag creation with autocompletion

mahezsh opened the pull request:
This PR Fixes #12509 

## Description

This shows the last 3 tags within the **Create Tag** component. 
It also filters the shown tags as the user type the tag name. (as suggested by @sergiou87)

### Screenshots

![screen shot](https://user-images.githubusercontent.com/32475101/129573665-dcb21b0f-ae0a-4a34-8149-b9a048ef3f93.gif)

<!--
If this PR touches the UI layer of the app, please include screenshots or animated gifs to show the changes.
-->

## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes: no-notes

mahezsh committed:
Enhancing tag creation with autocompletion

sergiou87 commented:
Thank you @matrixnl!! We love it 😍 

The only thing that feels jarring is when there are no matches and the whole `Previous Tags` section just disappears. We think it might be nice to keep the `Previous Tags` section with something like `No matches found for '7'` or something similar.

(Aside from that, I still have to review the code itself, I'll give it a first pass tomorrow 😄 )

sergiou87 commented on app/src/ui/create-tag/create-tag-dialog.tsx:
```diff
@@ -67,6 +71,17 @@ export class CreateTag extends React.Component<
             onValueChange={this.updateTagName}
           />
+
+          {this.state.previousTags && (
```
Nit, we try to avoid falsy/truthy checks
```suggestion
          {this.state.previousTags !== null && (
```

sergiou87 commented on app/src/ui/create-tag/create-tag-dialog.tsx:
```diff
@@ -99,9 +114,21 @@ export class CreateTag extends React.Component<
     return null
   }
 
+  private getExistingTagsFiltered(filter: string = ''): Array<string> | null {
+    const previousTags = []
+    if (this.props.localTags) {
+      for (const item of this.props.localTags.keys()) {
+        previousTags.push(item)
+      }
+      return previousTags.filter(item => item.includes(filter))
+    }
+    return null
+  }
+
```
We prefer checks + early returns, and also I think you can create an array from `localTags.keys()` just using `Array.from`

```suggestion
  private getExistingTagsFiltered(filter: string = ''): Array<string> | null {
    if (this.props.localTags === null) {
      return null
    }
    const previousTags = Array.from(this.props.localTags.keys())
    return previousTags.filter(item => item.includes(filter))
  }

```

sergiou87 reviewed (changes requested):
Looks good overall! Only a few suggestions 😄

mahezsh committed:
Handling no matches found

mahezsh commented:
Thanks @sergiou87, added **No matches found** and updated according to the suggestions 🙂

<img width="408" alt="Screen Shot" src="https://user-images.githubusercontent.com/32475101/129826970-f0eaddad-8ef5-42a1-bea3-d5a6d44206fb.png">

sergiou87 reviewed (approved):
I love it, thank you!!! ❤️

sergiou87 commented on app/src/ui/create-tag/create-tag-dialog.tsx:
```diff
@@ -67,6 +71,23 @@ export class CreateTag extends React.Component<
+              <p>Previous Tags</p>
+              {lastThreeTags.length === 0 ? (
+                <>
+                  <Ref>{`No matches found for '${this.state.tagName}'`}</Ref>
```
Just noticed in your screenshot (thanks for sharing that) something I overlooked while reviewing it.

I think for this specific scenario, we don't want to use a `Ref`, just normal text 😅 

Sorry about that!

mahezsh commented:
Created a new PR for this issue at #12966. Have cherry-picked the first 2 commits and addressed @sergiou87's comment.

sergiou87 commented:
Thanks, @mahezsh! I don't really understand why creating a new PR, but it's ok 😄  I guess we can close this one.
